### PR TITLE
Improve redis Swagger. 

### LINKED ISSUE #2968: Redis provider swagger describes GET patchSchedules by name but not LIST all
In fact the provider supports both, it was just forgotten from documentation, this is a pretty simple fix.

### PR DESCRIPTION
**Pull request by TimLovellSmith**
Documenting list all patchSchedules to fix #2968. Documenting the CheckNameAvailability.type property better to fix #2967. And using a realistic timespan value in PatchSchedule examples.

This checklist is used to make sure that common issues in a pull request are addressed. This will expedite the process of getting your pull request merged and avoid extra work on your part to fix issues discovered during the review process.

### PR information
- [x] The title of the PR is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For information on cleaning up the commits in your pull request, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).
- [x] Except for special cases involving multiple contributors, the PR is started from a fork of the main repository, not a branch.
- [x] If applicable, the PR references the bug/issue that it fixes.
- [x] Swagger files are correctly named (e.g. the `api-version` in the path should match the `api-version` in the spec).

### Quality of Swagger
- [x] I have read the [contribution guidelines](https://github.com/Azure/azure-rest-api-specs/blob/master/.github/CONTRIBUTING.md).
- [x] My spec meets the review criteria:
  - [x] The spec conforms to the [Swagger 2.0 specification](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md).
  - [x] The spec follows the guidelines described in the [Swagger checklist](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md) document.
  - [x] [Validation tools](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md#validation-tools-for-swagger-checklist) were run on swagger spec(s) and have all been fixed in this PR.

/cc @SiddharthChatrolaMs

### CODE REVIEW

**Comment by TimLovellSmith** [comment]
Autorest node generator appears to be crashing:

```
INFO:swaggertosdk.autorest_tools:   Loading AutoRest core      '/root/.autorest/@microsoft.azure_autorest-core@2.0.4275/node_modules/@microsoft.azure/autorest-core/dist' (2.0.4275)

INFO:swaggertosdk.autorest_tools:   Installing AutoRest extension '@microsoft.azure/autorest.nodejs' (^2.1.43)

INFO:swaggertosdk.autorest_tools:   Installing AutoRest extension '@microsoft.azure/autorest.modeler' (2.3.44)

INFO:swaggertosdk.autorest_tools:FATAL: System.InvalidOperationException: Sequence contains no matching element

INFO:swaggertosdk.autorest_tools:   at System.Linq.Enumerable.First[TSource](IEnumerable1 source, Func2 predicate)

INFO:swaggertosdk.autorest_tools:   at AutoRest.Extensions.SwaggerExtensions.RemoveUnreferencedTypes(CodeModel codeModel, HashSet`1 typeNames)

INFO:swaggertosdk.autorest_tools:   at AutoRest.Extensions.SwaggerExtensions.FlattenModels(CodeModel codeModel)

INFO:swaggertosdk.autorest_tools:   at AutoRest.Extensions.Azure.AzureExtensions.NormalizeAzureClientModel(CodeModel codeModel)

INFO:swaggertosdk.autorest_tools:   at AutoRest.NodeJS.Azure.TransformerJsa.AutoRest.Core.ITransformer<AutoRest.NodeJS.Azure.Model.CodeModelJsa>.TransformCodeModel(CodeModel cm) in /opt/vsts/work/1/s/src/azure/TransformerJsa.cs:line 34

INFO:swaggertosdk.autorest_tools:   at AutoRest.NodeJS.Program.<ProcessInternal>d__3.MoveNext() in /opt/vsts/work/1/s/src/Program.cs:line 125

INFO:swaggertosdk.autorest_tools:--- End of stack trace from previous location where exception was thrown ---

INFO:swaggertosdk.autorest_tools:   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()

INFO:swaggertosdk.autorest_tools:   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)

INFO:swaggertosdk.autorest_tools:   at NewPlugin.<Process>d__15.MoveNext()

INFO:swaggertosdk.autorest_tools:FATAL: nodejs/generate - FAILED

INFO:swaggertosdk.autorest_tools:FATAL: Error: Plugin nodejs reported failure.

INFO:swaggertosdk.autorest_tools:Process() cancelled due to exception : Plugin nodejs reported failure.

INFO:swaggertosdk.autorest_tools:  Error: Plugin nodejs reported failure.
```

File: specification/redis/resource-manager/Microsoft.Cache/stable/2018-03-01/redis.json [hunk 1]
```diff
@@ -1994,15 +2060,22 @@
...
       "properties": {
         "name": {
           "type": "string",
           "description": "Resource name."
         },
         "type": {
           "type": "string",
-          "description": "Resource type."
+          "description": "Resource type to check name availability of, e.g. 'Microsoft.Cache/redis'.",
+          "enum": [
```
[the remark this answers is not in the document]
Comment (TimLovellSmith) [reply specification/redis/resource-manager/Microsoft.Cache/stable/2018-03-01/redis.json]: @jhendrixMSFT What say you?

[the remark this answers is not in the document]
Comment (jhendrixMSFT) [reply specification/redis/resource-manager/Microsoft.Cache/stable/2018-03-01/redis.json]: I'm torn on this as breaking changes stink however in this case there was no clue what to pass for the value (I had to look at the checked in example JSON to figure it out).  If we really don't want to take the breaking change then the doc string will have to be the source of truth.
Another thing to consider, if this param will only ever have the value "Microsoft.Cache/redis" then if you mark it ```"modelAsString": false``` AutoRest should treat it as a constant and remove the param entirely (i.e. hard-coding the value in the request).

[the remark this answers is not in the document]
Comment (TimLovellSmith) [reply specification/redis/resource-manager/Microsoft.Cache/stable/2018-03-01/redis.json]: @jhendrixMSFT 
For this api-version we don't have any other resource types, and if we ever have a new resource we can use a new api-version for it. So I think it's fine to treat it as a constant.
Let's try that out and see if that makes everyone happy!
/cc @jianghaolu

**Comment by TimLovellSmith** [comment]
I see linter validation errors..
I think it is just because the parameter got removed and I need to remove it from the example too.
Trying that now..

`
error: Found errors in validating the request for x-ms-example "RedisCacheList" in operation "Redis_CheckNameAvailability".:

 

{ code: 'REQUEST_VALIDATION_ERROR',

  id: 'OAV109',

  message: 'Found errors in validating the request for x-ms-example "RedisCacheList" in operation "Redis_CheckNameAvailability".',

  innerErrors: 

   [ { code: 'INVALID_REQUEST_PARAMETER',

       errors: 

        [ { code: 'ONE_OF_MISSING',

            params: [],

            message: 'Data does not match any schemas from \'oneOf\'',

            path: [ 'type' ],

            inner: 

             [ { code: 'ENUM_MISMATCH',

                 params: [ 'Microsoft.Cache/Redis' ],

                 message: 'No enum match for: Microsoft.Cache/Redis',

                 path: [ 'type' ],

                 description: 'Resource type to check name availability of, e.g. \'Microsoft.Cache/redis\'.' },

               { code: 'INVALID_TYPE',

                 params: [ 'null', 'string' ],

                 message: 'Expected type null but found type string',

                 path: [ 'type' ] } ] } ],

       in: 'body',

       message: 'Invalid parameter (parameters): Value failed JSON Schema validation',

       name: 'parameters',

       path: 

        [ 'paths',

          '/subscriptions/{subscriptionId}/providers/Microsoft.Cache/CheckNameAvailability',

          'post',

          'parameters',

          '0' ] } ] }

{ AssertionError [ERR_ASSERTION]: swagger "specification/redis/resource-manager/Microsoft.Cache/stable/2018-03-01/redis.json" contains model validation errors.

    at Context.<anonymous> (/home/travis/build/Azure/azure-rest-api-specs/test/model.js:20:16)

    at <anonymous>

  generatedMessage: false,

  name: 'AssertionError [ERR_ASSERTION]',

  code: 'ERR_ASSERTION',

  actual: false,

  expected: true,

  operator: '==' }


  1 passing (3s)

  1 failing


  1) Azure swagger model validation using x-ms-examples and examples in spec

       specification/redis/resource-manager/Microsoft.Cache/stable/2018-03-01/redis.json should have valid examples.:


      AssertionError [ERR_ASSERTION]: swagger "specification/redis/resource-manager/Microsoft.Cache/stable/2018-03-01/redis.json" contains model validation errors.

      + expected - actual


      -false

      +true

      

      at Context.<anonymous> (test/model.js:20:16)

      at <anonymous>
`

**Comment by TimLovellSmith** [comment]
@jhendrixMSFT @jianghaolu 
Looks like removing the parameter and adding the enum type still gets flagged as a breaking change by AutoRest. Do you think that this is really a good idea?

If it were possible, I would rather do this in a non-breaking way (i.e. keep it string) but with a 'default' parameter value supplied, so that anyone writing new code wouldn't have to feel confused about what type to supply. Do you know if that is possible?

**Comment by jhendrixMSFT** [comment]
The question of default value has come up before and I don't believe there is support for this.  So if we want to do it in a non-breaking way the only real option that I know of is to update the comment like you already did.
CC @fearthecowboy

**Comment by TimLovellSmith** [comment]
@fearthecowboy @jhendrixMSFT 
I suppose you could have an 'x-ms-recommended-value' attribute for specifying a 'client side default' for code generators to use when the server side doesn't actually implementing a 'default value' behavior (at least, not per the Swagger 3.0 meaning of 'default'...) but it doesn't seem to make very good sense from a long-term point of view... since a 'default resource type' only makes sense until suddenly there's a new resource type, and then it really makes more sense to ask the client to specify...
However...
Even if we had another resource type which was not 'redis' cache, then we would probably encode it in a separate resource.json file, and we could just define it having a separate checkNameAvailability operation with a separate 'single enum value' so that even though they call the same server API they look like separate unrelated APIs to the codegen/client...
So I guess in this case it is not that crazy to either go with the single value enum (and breaking change) or really codegen a string parameter with a single default value for backwards compatibility... now and forever.

**Comment by TimLovellSmith** [comment]
@jhendrixMSFT  Given that we have other issues with CheckNameAvailability, which might be another breaking change, and a new api-version to fix (I refer to #2981) I think the best thing to do is NOT do any breaking change to the generated code for this api-version, but just document it for this api version and CONSIDER modeling it as an enum next api-version, and then people will hopefully notice the return type is changing at the same time as they are fixing up build issues for the number of arguments change.

So, for this PR, I am just going to improve the documentation, and revert the model as enum changes. I hope this makes sense.

/cc @fearthecowboy @hovsepm @jianghaolu

**Comment by TimLovellSmith** [comment]
@jhendrixMSFT @jianghaolu Is this PR good enough now?

**Review by jhendrixMSFT — APPROVED** [review]

**Comment by TimLovellSmith** [comment]
@jhendrixMSFT @jianghaolu Thanks for the review!

**Pull request merged**